### PR TITLE
Coverage report fixes

### PR DESCRIPTION
**Pull request by aescolar**
Several fixes to avoid coverage reports from changing in mysterious ways in between runs:
* Autostop ztests once they pass or fail to avoid sanitycheck's SIGTERM racing possible code which will continue executing after.
* We post-process the gcov output with lcov and feed that to codecov (instead of the raw gcov files), and we avoid feeding the gcovr output meant for shipabble.
* Excluded from the coverage reports the code in ARCH_POSIX which is not deterministic (what is executed depends on exactly when SIGTERM arrives).
* Excluded from the coverage reports some of the ARCH_POSIX code which is only there to handle error responses from the host (and will therefore never be covered).

Should fix the worst of #5938 
Merging #5774 will help both to increase the reported coverage and to get the coverage make some more sense in quite many places.
Solving #5723 and using it from sanitycheck to set CONFIG_COVERAGE for `samples/` will help in the same way as #5774 does already for `tests/`.

If you want to reproduce it quickly locally:
```
sanitycheck -p native_posix
rm lcov.*info
lcov --capture --directory sanity-out/native_posix/ --output-file lcov.pre.info -q --rc lcov_branch_coverage=1
lcov -q --remove lcov.pre.info *generated* -o lcov.info -rc lcov_branch_coverage=1
genhtml lcov.info --output-directory lcov_html -q --ignore-errors source  --branch-coverage --highlight --legend 
x-www-browser ./lcov_html/index.html &> /dev/null &
```

### CODE REVIEW

File: arch/posix/core/posix_core.c [hunk 1]
```diff
@@ -283,10 +283,12 @@ static void *posix_thread_starter(void *arg)
 	/*
 	 * The program may have been finished before this thread ever got to run
 	 */
+	/* LCOV_EXCL_START */ /* See Note1 */
```
Comment (pfalcon) [inline arch/posix/core/posix_core.c]: That's interesting pragma. Will it mark the code inside as not covered, or as "not subject to coverage" ? I.e. would such blocks preclude achieving 100% of coverage?

Comment (aescolar-ot) [reply arch/posix/core/posix_core.c]: I can only refer you to the lcov documentation: http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php

gcovr was supposed to support this pseudo pragmas too.

Comment (nashif) [reply arch/posix/core/posix_core.c]: the coverage reports we have and shown above do not use gcovr. gcovr is only used with shippable.

**Review by dbkinder — APPROVED** [review]
LGTM

**Comment by aescolar-ot** [comment]
@nashif : Do you like how this is looking?
If I feed codecov lcov's output instead of the raw gcov data things are easier: ac7320f51a1147123b2e8ee218d585944325ef0c
&
#5774 : helps to get the coverage make some more sense in quite many places

**Comment by nashif** [comment]
> Autostop ztests once they pass or fail to avoid sanitycheck's SIGTERM racing possible code which will continue executing after.

I am not really following how is this relevant. the SIGTERM comes after a test has completed execution, so whether you kill the app or auto-stop it it has exactly the same effect, because the code in question has already been executed and coverage data was generated.

**Comment by aescolar-ot** [comment]
>  the SIGTERM comes after a test has completed execution

In some tests, when the test passes or fails there is still code that will execute (non test threads may have been left over)
For example in tests/kernel/obj_tracing, 5 philosopher threads are left running after the test finishes.
sanitycheck is looking into the test stdout, and it will kill (SIGTERM) the test as soon as it sees the ``PROJECT EXECUTION *`` line.
So, in this case, 
https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/kernel/obj_tracing/src/philosopher.c#L55 

You have sanitycheck's SIGTERM racing these threads code. Depending on the relative delay of SIGTERM, more or less code will be executed in the philosophers threads, and you will see changes in the coverage.

You can see here a trace with timestamps of this testcase, where I added an extra print next to the last line:
https://pastebin.com/XjeHyvrQ
(note that the times are simulated times, so it all happens in a blink)
You can see that the test thread finishes in 0 time (before the first time the cpu goes to sleep), and then the philosophers continue.

Here one example run where no thread reached the `GIVE(&f3)`:
https://codecov.io/gh/zephyrproject-rtos/zephyr/compare/2151b864d4750e9b476cbc140b6c4baff59b7296...ada23b3cdaf9879cb09321590670486a15c3ef6d/src/tests/kernel/obj_tracing/src/philosopher.c

There is also example runs of this same test where no philosopher reaches `GIVE(&f2)` either.

**Comment by aescolar-ot** [comment]
There is a small issue with what we are feeding to codecov.io.
Depending on what we feed we will get quite different results.

### Raw gcov data + gcovr output [Current master]:
Right  now in the master branch, we are actually feeding both the raw gcov files, and the xml file produced by gcovr meant for shippable.
See for example:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/10111/1/console
https://codecov.io/gh/zephyrproject-rtos/zephyr/commit/54d7207e09626470cd912a584fb22cfc28be0641/build

There is 3 problems with this:
1.	Both gcov and lcov understand LCOV_EXCL* pragmas, but codecov does not handle them on its side properly when it processes gcov raw data.
2.	Branch coverage is not properly reported.
3.	For some reason some files (e.g. drivers/sensor) which are never compiled in, are part of the report. But not all of them (e.g. drivers/rtc).


### gcovr output + lcov output:
If I remove the raw gcov output and instead generate with lcov an aggregated report:
1.	✔ LCOV_EXCL* is properly handled. 
2.	❌ Branch coverage is not properly handled (each report alone seems to have branch coverage. But maybe how codecov combines them is losing part of that info)
3.	❌ Spurious files are still there.

https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/10124/1/console 
```
==> Reading reports
 + ./lcov.info bytes=627658
 + ./shippable/codecoverage/coverage.xml bytes=3783614
```
https://codecov.io/gh/zephyrproject-rtos/zephyr/commit/da80231a076557734d624320f320dc674d691da0/build 


### gcovr output alone:
Feeding it only gcovr’s output:
1.	✔ LCOV_EXCL* is properly handled. 
2.	✔ Branch coverage is properly handled => the reported coverage drops (-6%)
3.	❌ Spurious files are still there.
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/10112/1/console
https://codecov.io/gh/zephyrproject-rtos/zephyr/commit/689be53b4347f0c7ded8c980882f31ced31f5197/build


### lcov output alone: 
Feeding it only lcov’s output:
1.	✔ Branch coverage is properly handled => the reported coverage drops (-6%)
2.	✔ LCOV_EXCL* is properly handled. 
3.	✔ No spurious files (=> significant increase in reported % coverage)

https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/10129/1/console
https://codecov.io/github/zephyrproject-rtos/zephyr/commit/9899951bce0b5faac4e58ee38347515889e00f8c


I lean to only feed lcov’s report to codecov. so, I modified this PR to do so.

**Review by nashif — APPROVED** [review]

**Pull request merged**